### PR TITLE
Redundancy analysis

### PR DESCRIPTION
**Pull request by MikeLippincott**
This PR performs a redundancy analysis using Canonical Correlation Analysis coefficients to calculate a Redundancy index for a CCA latent space. These are then used to determine the degree of which one of the two paired datasets are redundant.

### CODE REVIEW

File: 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb [hunk 1]
```diff
@@ -0,0 +1,1098 @@
+{
```
Comment (MattsonCam) [inline 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.        nomic_data.columns[nomic_data.columns.str.contains("[NSU]", regex=True)]
</pre><p>May consider setting this to a variable so you don't have this long expression each time</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/WayScience/pyroptosis_signature_data_analysis/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (MattsonCam) [inline 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    # shuffle the data both rows and columns
</pre><p>Does this also shuffle the columns? It looks like it shuffles the row values for each column?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/WayScience/pyroptosis_signature_data_analysis/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (MattsonCam) [inline 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>Is each sample a cell here, so each of the rows correspond to a sample, and the same sample would match in both the morphology data and the nomic data?</p><p> </p><p>Also, what are the canonical variables in this case (i.e. how are they defined)?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/WayScience/pyroptosis_signature_data_analysis/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (MattsonCam) [inline 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    N = morphology_data.shape[0]
</pre><p>I think you define N twice here. If you want to show they are equal consider adding a print or assert statement</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/WayScience/pyroptosis_signature_data_analysis/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (MattsonCam) [inline 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>May consider making these equations larger to better see the ~</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/WayScience/pyroptosis_signature_data_analysis/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (MattsonCam) [inline 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>Slight typo spelling sklearn</p><p> </p><p>I like how you included references</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/WayScience/pyroptosis_signature_data_analysis/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (MattsonCam) [inline 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.    K = min(N, P, Q)
</pre><p>Consider checking if N &lt; min(P, Q) first</p><p> </p><p>Why do K depend on the number of rows N?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/WayScience/pyroptosis_signature_data_analysis/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (MattsonCam) [inline 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>Are the a and b terms here supposed to be (x_p - x_mu)^2 and (x_q - x_mu)^2 for the variance? May consider changing this below if this is the case</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/WayScience/pyroptosis_signature_data_analysis/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (MattsonCam) [inline 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>I'm confused, I thought this is supposed to be between 0 and 1?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/WayScience/pyroptosis_signature_data_analysis/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (MattsonCam) [inline 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>Slight typo in first line</p><p> </p><p>Could also simplify to RI_total = sum(from k=1 to K of RI_u + RI_v)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/WayScience/pyroptosis_signature_data_analysis/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (MattsonCam) [inline 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>Are the min and max of the redundancy indices valuable info?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/WayScience/pyroptosis_signature_data_analysis/pull/75/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

**Review by MattsonCam — APPROVED** [review]
Hey @MikeLippincott there may be a few changes to be addressed, but overall, good job!

File: 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb (hunk 1 above)
Comment (MikeLippincott) [reply 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>Good catch:</p><p>Change in documentation:</p><p># shuffle the data for each column</p><p># this changes the single well values but keeps the distribution of each column</p>

Comment (MikeLippincott) [reply 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>each sample is a well in this case. I updated the notebook with documentation to match.</p><p> </p><p>The canonical variables are defined later. I also added documentation to reflect this</p>

Comment (MikeLippincott) [reply 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>Yes, thank you. I removed that line now.</p>

Comment (MikeLippincott) [reply 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>Yes, good suggestion. I added the size increase.</p>

Comment (MikeLippincott) [reply 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>I do not think I should need to check as if N is indeed &lt; the min(P,Q) then it will be picked up in the expression min(N,P,Q) but if it is larger then it wont be used.</p><p> </p><p>In terms of why - this is a great questions that the answer is not documented and I added documentation to address this. Thank you for checking in on this!</p>

Comment (MikeLippincott) [reply 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>I do not think so but let us talk with Greg more on Monday to find out. </p>

Comment (MikeLippincott) [reply 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>Me, too...</p><p> </p><p>But sklearn uses a difference definition. It is the <code>R2_adjusted</code> term </p><p>from the website: "This is not a symmetric function.</p><p>Unlike most other scores, </p><p>score may be negative (it need not actually be the square of a quantity R)."</p><p> </p><p><a href="https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html" target="_blank">https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html</a></p><p> </p><p>Blew my mind too...</p>

Comment (MikeLippincott) [reply 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>Adjusted! Thank you for pointing out</p>

Comment (MikeLippincott) [reply 7.redundancy_analysis/notebooks/0.redundancy_analysis.ipynb:1]: <p>These were used for plotting to keep a square scale when visualizing the results</p>
